nullaway/src/main/java/com/uber/nullaway/handlers/ContractHandler.java

```diff
@@ -0,0 +1,192 @@
+      MethodInvocationNode node,
+      Types types,
+      AccessPathNullnessPropagation.SubNodeValues inputs,
+      AccessPathNullnessPropagation.Updates thenUpdates,
+      AccessPathNullnessPropagation.Updates elseUpdates,
+      AccessPathNullnessPropagation.Updates bothUpdates) {
+    Symbol.MethodSymbol callee = ASTHelpers.getSymbol(node.getTree());
+    Preconditions.checkNotNull(callee);
+    // Check to see if this method has an @Contract annotation
+    if (ASTHelpers.hasDirectAnnotationWithSimpleName(callee, "Contract")) {
+      // Found a contract, lets parse it.
+      String[] clauses = getContractFromAnnotation(callee).split(";");
+      for (String clause : clauses) {
+        String[] parts = clause.split("->");
+        assert parts.length == 2;
+        String[] antecedent = parts[0].split(",");
+        String consequent = parts[1].trim();
+        // Find a single value constraint that is not already known. If more than one arguments with
+        // unknown
+        // nullness affect the method's result, then ignore this clause.
+        int argIdx = -1;
+        Nullness argAntecedentNullness = null;
+        boolean supported =
+            true; // Set to false if the rule is detected to be one we don't yet support
+        // Fixme: Proper error reporting for malformed clauses:
+        assert antecedent.length == node.getArguments().size();
+        for (int i = 0; i < antecedent.length; ++i) {
+          String valueConstraint = antecedent[i].trim();
+          if (valueConstraint.equals("_")) {
+            continue;
+          } else if (valueConstraint.equals("false") || valueConstraint.equals("true")) {
+            supported = false;
+            break;
+          } else if (valueConstraint.equals("!null")
+              && inputs.valueOfSubNode(node.getArgument(i)).equals(Nullness.NONNULL)) {
+            // We already know this argument can't be null, so we can treat it as not part of the
+            // clause
+            // for the purpose of deciding the non-nullness of the other arguments.
+            continue;
+          } else if (valueConstraint.equals("null") || valueConstraint.equals("!null")) {
+            if (argIdx != -1) {
+              // More than one argument involved in the antecedent, ignore this rule
+              supported = false;
+              break;
+            }
+            argIdx = i;
+            argAntecedentNullness =
+                valueConstraint.equals("null") ? Nullness.NULLABLE : Nullness.NONNULL;
+          } else {
+            assert false;
+            supported = false;
+            break;
+          }
+        }
+        if (!supported) {
+          // Too many arguments involved, or unsupported @Contract features. On to next clause in
+          // the
+          // contract expression
+          continue;
+        }
+        if (argIdx == -1) {
+          // The antecedent is unconditionally true. Check for the ... -> !null case and set the
```

Comment:
Don't fully know the antecedent language, but are we sure it is equivalent to `true` here?  `@Contract` can't express any conditions on parameters that we don't understand?

Reply:
Nope, and yup, respectively :) 

This is the contract language, we handle all `value-constraint` cases **and** also set `supported = false`, which causes us to bail before this point, if we ever see any arg in the clause that doesn't match those five cases.

```
contract ::= (clause ‘;’)* clause
clause ::= args ‘->’ effect
args ::= ((arg ‘,’)* arg )?
arg ::= value-constraint
value-constraint ::= ‘_’ | ‘null’ | ‘!null’ | ‘false’ | ‘true’
effect ::= value-constraint | ‘fail’
```

Reply:
Actually, I wonder if there is a parser generator lightweight enough for us to actually use here, rather than doing it the hacky-ish way. Since they were nice enough to give us a grammar... know any?

Reply:
Oh I see, the key is to bail when `supported = false`; got it.  Re: the parser generator, probably not worth it; I don't know of anything super lightweight.